Implement fix reorg and remove gfs-utils code

WalterKolczynski-NOAA opened the pull request:
**Description**
Removes all of the code and scripts that were moved to the new [gfs-utils repo](https://github.com/NOAA-EMC/gfs-utils) and adjusts workflow scripts to build and use them from the new location.

Some of the build scripts had unnecessary calls to machine-setup that are removed because the lower-level script already has the same functionality.

This PR also includes updates to use the new fix organization. This includes the addition of a fix versions file, updates to the link script, and some changes in the fix directories used by scripts to account for files that have been relocated. The versions file sets the version number for each component of fix, so that fix can be more easily maintained and documented. The initial versions are all the same, and correspond to the old fix_NEW directory (other than some directories have been renamed or reorganized). Except gdas, which has already had a new set of fix files added. The fix update also required an update to UFS-Utils.

Finally, the link script has been updated to match the syntax of the rest of the build system ([checkout.sh](sorc/checkout.sh) and [build_all.sh](sorc/build_all.sh)). [link_workflow.sh](sorc/link_workflow.sh) now detects the machine automatically instead of requiring an argument, and dev mode, which used to be set using `emc` as an argument, is now assumed. To run in ops mode (copy instead of link), the `-o` option is used. The full syntax is now simply:
```
./link_workflow.sh [-o]

Fixes #356
Fixes #966
Depends on ufs-community/UFS_UTILS#688

**Type of change**
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**How Has This Been Tested?**
- [x] 5½-cycle cycling test on Hera
- [x] Coupled test on Hera
- [x] 2½-cycle cycling test on Orion
- [x] Coupled test on Orion

**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

KateFriedman-NOAA committed:
Update fix subfolders to ignore in .gitignore

Refs: #966

KateFriedman-NOAA committed:
Update link_workflow.sh for fix set reorg

- Source new fix.ver file to get fix subfolder version values for setting up fix folder symlinks.
- Change "fix_NEW" to "fix" for FIX_DIR paths: use newly reorganized and versioning fix set.
- Remove "fix_" from /fix folder symlink names.
- Update /fix folder symlink creation to use version variables from fix.ver.
- Update GSI and GDAS fix symlinks.

Refs: #966

KateFriedman-NOAA committed:
Update JEDI fix subfolder path name in configs

- Change FV3JEDI_FIX in related configs to change the path subfolder from
"fix_jedi" to "gdas".

Refs: #966

KateFriedman-NOAA committed:
Update fix subfolders in configs to remove "fix_"

- Remove "fix_" text from fix file paths in several configs.

Refs: #966

KateFriedman-NOAA committed:
Remove "fix_" from fix subfolder paths in scripts

- Update /scripts and /ush scripts to remove the "fix_" text
from fix subfolder paths.

Refs: #966

KateFriedman-NOAA committed:
Create versions folder and fix.ver

- Create a new version folder to hold version files.
- Create a new fix.ver to hold fix file versions.

Refs: #966

WalterKolczynski-NOAA committed:
Remove gfs-utils code and update for new fix

Removes all of the code and scripts that were moved to the new gfs-utils
repo and adjusts workflow scripts to build and use them from the new
location.

Some of the build scripts had unnecessary calls to machine-setup that are
removed because the lower-level script already has the same functionality.

This PR also includes updates to use the new fix organization. This includes
the addition of a fix versions file, updates to the link script, and some
changes in the fix directories used by scripts to account for files that
have been relocated. The versions file sets the version number for each
component of fix, so that fix can be more easily maintained and documented.
The initial versions are all the same, and correspond to the old fix_NEW
directory (other than some directories have been renamed or reorganizated).

Finally, the link script has been updated to match the syntax of the rest
of the build system (`checkout.sh` and `build_all.sh`). `link_workflow.sh`
now detects the machine automatically instead of requiring an argument, and
dev mode, which used to be set using `emc` as an argument, is now assumed.
To run in ops mode (copy instead of link), the `-o` option is used. The
full syntax is now simply:
```
./link_workflow.sh [-o]

-o: Run in operations (NCO) mode (copy instead of link)
```

This commit uses a temporary clone of the UFS_UTILS repo pending the merger
of ufs-community/UFS_UTILS#688.

Fixes #356

WalterKolczynski-NOAA commented:
Currently using a temporary repo for UFS_UTILS until ufs-community/UFS_UTILS#688 is merged.

WalterKolczynski-NOAA committed:
Correct merge error in link script

An incorrect resolution of a merge conflict resulted in an undefined
variable in the link script.

Refs: #356

WalterKolczynski-NOAA committed:
Update gfs utils to pick up Orion gempak fix

Updates the gfs utils version to pick up a fix for an incorrect gempak
module version on Orion.

WalterKolczynski-NOAA committed:
Update to new UFS_UTILS hash

WalterKolczynski-NOAA committed:
Revert change to checkout for GSI-adjacent repos

An earlier change had inadvertently stopped checking out GSI Monitor,
GSI Utils, and GLDAS when GDAS is used instead of GSI.

Refs: #966

WalterKolczynski-NOAA committed:
Add gfs-utils to manage externals config

Refs: #966

WalterKolczynski-NOAA committed:
First wave of linter fixes

First wave of linter fixes for removing gfs-utils code/fix reorg.

Refs: #966

WalterKolczynski-NOAA committed:
Linter fixes for fix dir reorg

Second wave of linter fixes for removing gfs-utils code/fix reorg.

Refs: #966

WalterKolczynski-NOAA committed:
Wave 3 of linter fixes for fix update

Wave 3 of linter fixes for removing gfs-utils code/fix reorg.

Refs: #966

WalterKolczynski-NOAA committed:
Wave 4 linter fixes for fix update

Wave 4 of linter fixes for removing gfs-utils code/fix reorg.

Refs: #966

WalterKolczynski-NOAA commented:
Going to rerun a test or two over the weekend from a clean install to confirm things still work, but should be able to start reviewing.

WalterKolczynski-NOAA committed:
Fix minor syntax error in link script

The last merge from develop introduced a minor syntax error in the
link script due to a bad conflict resolution. The issue is now
corrected.

Refs: #966

WalterKolczynski-NOAA commented:
Re-tests pass

WalterKolczynski-NOAA committed:
Update new GLDAS scripts for new fix pattern

GLDAS scripts were recent moved into the workflow repo and need to
be updated for the new fix structure like other components.

Refs: #622, #966

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -1,372 +1,421 @@
+# shellcheck disable=SC1091
+source gfs_utils.fd/ush/machine-setup.sh > /dev/null 2>&1
+# shellcheck disable=
+machine="${target:?}"
```
Why do we need this line.  `machine` was detected from L47.  If it is not supported, this is a hard exit, no?

WalterKolczynski-NOAA replied:
[machine-setup.sh](https://github.com/NOAA-EMC/gfs-utils/blob/develop/ush/machine-setup.sh) sets `$target`, but this script for some reason uses `$machine`. I didn't want to make yet more changes to change them all in this PR.

WalterKolczynski-NOAA replied:
Actually, it is only used a couple places. I think I can change it.

WalterKolczynski-NOAA replied:
Though I'd rather machine-setup be setting `$machine` instead of `$target` in the first place.

WalterKolczynski-NOAA replied:
Going to leave this for now unless there is an objection. We can switch all scripts to use `$machine` in the future.

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -1,372 +1,421 @@
-if [ ! -z $FIX_DIR ]; then
- if [ ! -d ${pwd}/../fix ]; then mkdir ${pwd}/../fix; fi
+# Source fix version file
+source "${top_dir}/versions/fix.ver"
```
L51 sources this already. Why do it again?

WalterKolczynski-NOAA replied:
I think this was an error when I cherry-picked Kate's changes over. I'll remove it.

WalterKolczynski-NOAA replied:
Redundant load removed

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -1,372 +1,421 @@
+  "hera")     FIX_DIR="/scratch1/NCEPDEV/global/glopara/fix" ;;
+  "orion")    FIX_DIR="/work/noaa/global/glopara/fix" ;;
+  "jet")      FIX_DIR="/lfs4/HFIP/hfv3gfs/glopara/git/fv3gfs/fix" ;;
+  *)
+    echo "FATAL: Unknown target machine ${machine}, couldn't set FIX_DIR"
+    exit 1
```
This should be check upfront at the earliest possible time available.

WalterKolczynski-NOAA replied:
Agree, not sure how it got moved this far away from machine-setup.

WalterKolczynski-NOAA replied:
Moved to directly after the machine-setup, which is the first thing after option parsing.

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -1,372 +1,421 @@
-    echo ' Syntax: link_workflow.sh ( nco | emc ) ( wcoss2 | hera | orion | jet | stampede )'
-    exit 1
-fi
+script_dir=$(cd "$(dirname "${BASH_SOURCE[0]}")" &> /dev/null && pwd)
```
Isn't this `sorc_dir`?

WalterKolczynski-NOAA replied:
It would be the sorc dir, yes.

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -1,372 +1,421 @@
-    exit 1
-fi
+script_dir=$(cd "$(dirname "${BASH_SOURCE[0]}")" &> /dev/null && pwd)
+top_dir=$(cd "$(dirname "${script_dir}")" &> /dev/null && pwd)
```
This is `HOMEgfs`, no?

WalterKolczynski-NOAA replied:
Yes, but that isn't a thing here. This isn't a runtime script.

aerorahul commented on sorc/link_workflow.sh:
```diff
@@ -1,372 +1,421 @@
     if [ -d gsi_enkf.fd ]; then
       [[ -d gsi.fd ]] && rm -rf gsi.fd
-      $SLINK gsi_enkf.fd/src/gsi                                                                gsi.fd
+      ${SLINK} gsi_enkf.fd/src/gsi                                                                gsi.fd
```
We are overloading `${LINK}` as either copy or link based on NCO or EMC mode.
Can we replace `${SLINK}` and call it `${NLN}`?  There is a chance someone will misunderstand and get confused between the interchangeable `LINK` and `SLINK`.

WalterKolczynski-NOAA replied:
I don't know that NLN is all that much clearer, but at any rate I wasn't going to change it for this PR. It has been `LINK` and `SLINK` for a long time.

KateFriedman-NOAA commented on ush/gfs_bufr.sh:
```diff
@@ -19,30 +19,18 @@
 # echo "History: February 2003 - First implementation of this utility script"
 #
-source "$HOMEgfs/ush/preamble.sh"
+source "${HOMEgfs:-?}/ush/preamble.sh"
```
Is there a difference between `:-?` and `:?`?

WalterKolczynski-NOAA replied:
Actually, `:?` is correct. I need to change these.

WalterKolczynski-NOAA replied:
Fixed

WalterKolczynski-NOAA committed:
Fix default missing var syntax error

Corrects a `:-?` to `:?`.

Refs: #356

aerorahul commented on ush/gfs_truncate_enkf.sh:
```diff
@@ -14,7 +14,7 @@ mkdir -p $DATATMP
 
 export LEVS=${LEVS_LORES:-64}
-export FIXam=${FIXam:-$HOMEgfs/fix/fix_am}
+export FIXam=${FIXam:-$HOMEgfs/fix/am}
```
I doubt this script is used today.  @yangfanglin do you know?

aerorahul commented on ush/fv3gfs_regrid_nemsio.sh:
```diff
@@ -28,9 +28,9 @@ DATA=${DATA:-$pwd}
-FIX_AM=${FIX_AM:-$FIX_DIR/fix_am}
-FIXfv3=${FIXfv3:-$FIX_DIR/fix_fv3_gmted2010}
-REGRID_NEMSIO_EXEC=${REGRID_NEMSIO_EXEC:-$HOMEgfs/exec/regrid_nemsio}
+FIX_AM=${FIX_AM:-$FIX_DIR/am}
```
This script can be deleted.  It is no longer in use.

WalterKolczynski-NOAA replied:
We have a separate issue already open to remove all the nemsio stuff. Someone just needs to go through it all.

aerorahul commented on ush/fv3gfs_remap.sh:
```diff
@@ -16,7 +16,7 @@ export DATA=${DATA:-$pwd}
 export HOMEgfs=${HOMEgfs:-$NWPROD}
 export FIX_DIR=${FIX_DIR:-$HOMEgfs/fix}
-export FIXfv3=${FIXfv3:-$FIX_DIR/fix_fv3_gmted2010}
+export FIXfv3=${FIXfv3:-$FIX_DIR/orog}
```
This script is not used in the workflow.
If it is necessary, it belongs in ufs-utils.  # TODO

aerorahul reviewed (commented):
looks good.  Some comments.
This PR would have been much easier to review if it were separate PRs:
1. removing gfs-utils
2. fix reorg
3. updates to build scripts
4. updates to `link_workflow.sh`

WalterKolczynski-NOAA committed:
Move root fixdir setting higher in link script

Moved the setting of FIXDIR higher so it errors on an unknown machine
sooner. Also removed redundant version file load.

Refs: #966

WalterKolczynski-NOAA committed:
Fix erronous CR in log directory

The closing quotation mark when making the logdir in checkout was
erroneously split into a new line.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/use_gw_utils
